spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/datamining/RowTransformer.scala

```diff
@@ -0,0 +1,324 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.dataset.datamining
+
+import com.intel.analytics.bigdl.dataset.Transformer
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric.{NumericBoolean, NumericDouble, NumericFloat, NumericInt, NumericLong, NumericShort, NumericString}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.{NumericWildcard, TensorNumeric}
+import com.intel.analytics.bigdl.utils.{T, Table}
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+/**
+ * RowTransformer transform a `Row` to a `Table` whose values are all `Tensor`.
+ * This transformer is a container of `RowTransformSchema`s.
+ * When this transformer being executed,
+ * it will run `transform` methods of its `RowTransformSchema`s.
+ *
+ * Output of `RowTransformer` is a `Table`.
+ * The keys of `Table` are `schemaKey`s of included `RowTransformSchema`s.
+ * Correspondingly, the values of `Table` are results(`Tensor`) of `RowTransformSchema.transform`.
+ *
+ * @param schemas schemas of transformer, whose keys should `NOT` be duplicated
+ * @param rowSize size of `Row` transformed by this transformer, default is `None`
+ */
+class RowTransformer(
+  @transient private val schemas: Seq[RowTransformSchema],
+  protected val rowSize: Option[Int] = None
+) extends Transformer[Row, Table] {
+
+  protected val schemaMap: mutable.Map[String, RowTransformSchema] = {
+    val map = mutable.LinkedHashMap[String, RowTransformSchema]()
+    schemas.foreach { schema =>
+      require(!map.contains(schema.schemaKey),
+        s"Found replicated schemeKey: ${schema.schemaKey}"
+      )
+      if (schema.fieldNames.isEmpty) {
+        require(schema.indices.forall(i => i >= 0 && i < rowSize.getOrElse(Int.MaxValue)),
+          s"At least one of indices are out of bound: ${schema.indices.mkString(",")}"
+        )
+      }
+      map += schema.schemaKey -> schema
+    }
+    map
+  }
+
+  override def apply(prev: Iterator[Row]): Iterator[Table] = {
+    new Iterator[Table] {
+      override def hasNext: Boolean = prev.hasNext
+
+      override def next(): Table = {
+        val row = prev.next()
+        val table = T()
+        schemaMap.foreach { case (key, schema) =>
+          val indices = schema match {
+            case sch if sch.fieldNames.nonEmpty =>
+              schema.fieldNames.map(row.fieldIndex)
+            case sch if sch.indices.nonEmpty =>
+              schema.indices
+            case _ =>
+              0 until row.length
+          }
+
+          val (values, fields) = indices.map(i =>
+            row.get(i) -> row.schema.fields(i)
+          ).unzip
+
+          val output = schema.transform(values, fields)
+
+          table.update(key, output)
```

Comment:
is the key String? I suggest that use Tensor.scalar[String](key) to wrap it as a Tensor[String]

Reply:
Okay!